File: modules/15-reasoning-test-time-compute/ts/03-self-refine.ts
/**
 * Task 3 — Self-refine / reflection 🟡
 *
 * What this teaches:
 *   - Draft → critique → revise is a three-turn loop that mimics how a human
 *     reviews their own work. Each iteration uses the model to judge AND improve
 *     its own previous output.
 *   - Even a single round of self-refinement frequently improves factual accuracy,
 *     completeness, and clarity on open-ended tasks.
 *   - Multiple iterations converge quickly: the marginal gain from iteration 3 is
 *     usually not worth iteration 2's cost. Measure, don't assume.
 *
 * How to run:
 *   pnpm tsx modules/15-reasoning-test-time-compute/ts/03-self-refine.ts
 */

import "dotenv/config";
import { getProvider, ChatMessage, ChatOptions } from "@learn-ai/llm-core";

// ---------------------------------------------------------------------------
// Tasks to refine — open-ended so there is real room for improvement.
// ---------------------------------------------------------------------------
const TASKS = [
  {
    id: "explain",
    prompt:
      "Explain how a transformer attention mechanism works to a software " +
      "engineer who has not studied machine learning. Use an analogy.",
  },
  {
    id: "plan",
    prompt:
      "Write a brief 5-step plan for launching a personal blog in 2025, " +
      "including tools, audience strategy, and monetisation.",
  },
];

const MAX_ITERATIONS = 2;

const CRITIQUE_SYSTEM =
  "You are a critical reviewer. Read the task and the draft answer below.\n" +
  "List 3–5 specific issues, gaps, or inaccuracies in the draft. " +
  "Be concrete — do not say 'it could be clearer'; say exactly what is unclear and why.\n" +
  "Format your response as a numbered list.";

const REVISE_SYSTEM =
  "You are a skilled writer. You will receive a task, a draft answer, and a critique.\n" +
  "Produce a revised answer that directly addresses every point in the critique. " +
  "Your revision should be complete and self-contained — do not reference the critique in your output.";

// ---------------------------------------------------------------------------
// TODO 1: Implement draft(taskPrompt).
//         Generate an initial answer to the task. Use temperature=0.7 so the
//         output is natural but not overly conservative.
//         Return the text of the draft.
// ---------------------------------------------------------------------------
async function draft(taskPrompt: string): Promise<string> {
  // const llm = getProvider();
  // const result = await llm.chat(
  //   [{ role: "user", content: taskPrompt }],
  //   { temperature: 0.7 },
  // );
  // return result.text;
  throw new Error("TODO: implement draft()");
}

// ---------------------------------------------------------------------------
// TODO 2: Implement critique(taskPrompt, draftText).
//         Send a message with the critique system prompt + the task + the draft.
//         Return the critique text.
//         Use temperature=0 to keep the critique factual and repeatable.
// ---------------------------------------------------------------------------
async function critique(taskPrompt: string, draftText: string): Promise<string> {
  // const llm = getProvider();
  // const result = await llm.chat(
  //   [
  //     { role: "system", content: CRITIQUE_SYSTEM },
  //     { role: "user", content: `Task: ${taskPrompt}\n\nDraft answer:\n${draftText}` },
  //   ],
  //   { temperature: 0 },
  // );
  // return result.text;
  throw new Error("TODO: implement critique()");
}

// ---------------------------------------------------------------------------
// TODO 3: Implement revise(taskPrompt, draftText, critiqueText).
//         Send a message with the revise system prompt + task + draft + critique.
//         Return the revised text.
// ---------------------------------------------------------------------------
async function revise(
  taskPrompt: string,
  draftText: string,
  critiqueText: string,
): Promise<string> {
  // const llm = getProvider();
  // const result = await llm.chat(
  //   [
  //     { role: "system", content: REVISE_SYSTEM },
  //     {
  //       role: "user",
  //       content:
  //         `Task: ${taskPrompt}\n\n` +
  //         `Draft answer:\n${draftText}\n\n` +
  //         `Critique:\n${critiqueText}`,
  //     },
  //   ],
  //   { temperature: 0.3 },
  // );
  // return result.text;
  throw new Error("TODO: implement revise()");
}

// ---------------------------------------------------------------------------
// TODO 4: Implement compare(original, final).
//         A simple heuristic: length change, unique new words, changed at all.
//         Return an object. Replace with an LLM-as-judge call if you like.
// ---------------------------------------------------------------------------
function compare(
  original: string,
  final: string,
): { lengthDelta: number; newUniqueWords: number; changed: boolean } {
  // TODO: implement
  const lengthDelta = final.length - original.length;
  const originalWords = new Set(original.toLowerCase().split(/\s+/));
  const finalWords = new Set(final.toLowerCase().split(/\s+/));
  const newUniqueWords = [...finalWords].filter((w) => !originalWords.has(w)).length;
  return { lengthDelta, newUniqueWords, changed: original.trim() !== final.trim() };
}

function divider(label: string): void {
  console.log(`\n${"=".repeat(10)} ${label} ${"=".repeat(10)}`);
}

async function main() {
  console.log("=== Task 3: Self-Refine / Reflection ===\n");

  for (const task of TASKS) {
    console.log(`\nTask [${task.id}]: ${task.prompt}\n`);

    // -------------------------------------------------------------------------
    // TODO 5: Run the draft → critique → revise loop.
    //         1. Generate initial draft.
    //         2. For each iteration: critique the current draft, then revise it.
    //         3. Print each step so the reader can see the improvement.
    // -------------------------------------------------------------------------

    try {
      divider("DRAFT");
      let current = await draft(task.prompt);
      const original = current;
      console.log(current);

      for (let i = 1; i <= MAX_ITERATIONS; i++) {
        divider(`CRITIQUE (iteration ${i})`);
        const crit = await critique(task.prompt, current);
        console.log(crit);

        divider(`REVISION (iteration ${i})`);
        current = await revise(task.prompt, current, crit);
        console.log(current);
      }

      // -------------------------------------------------------------------------
      // TODO 6: Print a comparison between the original draft and final revision.
      // -------------------------------------------------------------------------
      divider("COMPARISON");
      const stats = compare(original, current);
      console.log(`  Length change     : ${stats.lengthDelta >= 0 ? "+" : ""}${stats.lengthDelta} chars`);
      console.log(`  New unique words  : ${stats.newUniqueWords}`);
      console.log(`  Changed at all?   : ${stats.changed}`);

    } catch (e: unknown) {
      const msg = e instanceof Error ? e.message : String(e);
      console.log(`  ${msg}`);
    }

    console.log();
  }

  console.log(
    "Observation: self-refinement tends to add specificity and fix factual gaps.\n" +
    "Two iterations usually capture most of the gain."
  );
}

main().catch(console.error);

File: modules/15-reasoning-test-time-compute/py/03_self_refine.py
"""
Task 3 — Self-refine / reflection 🟡

What this teaches:
  - Draft → critique → revise is a three-turn loop that mimics how a human
    reviews their own work. Each iteration uses the model to judge AND improve
    its own previous output.
  - Even a single round of self-refinement frequently improves factual accuracy,
    completeness, and clarity on open-ended tasks.
  - Multiple iterations converge quickly: the marginal gain from iteration 3 is
    usually not worth iteration 2's cost. Measure, don't assume.

How to run:
  uv run python modules/15-reasoning-test-time-compute/py/03_self_refine.py
"""

from __future__ import annotations

from llm_core import get_provider, ChatMessage, ChatOptions

# ---------------------------------------------------------------------------
# Tasks to refine — open-ended so there is real room for improvement.
# ---------------------------------------------------------------------------
TASKS = [
    {
        "id": "explain",
        "prompt": (
            "Explain how a transformer attention mechanism works to a software "
            "engineer who has not studied machine learning. Use an analogy."
        ),
    },
    {
        "id": "plan",
        "prompt": (
            "Write a brief 5-step plan for launching a personal blog in 2025, "
            "including tools, audience strategy, and monetisation."
        ),
    },
]

MAX_ITERATIONS = 2  # number of critique-revise cycles

CRITIQUE_PROMPT = (
    "You are a critical reviewer. Read the task and the draft answer below.\n"
    "List 3–5 specific issues, gaps, or inaccuracies in the draft. "
    "Be concrete — do not say 'it could be clearer'; say exactly what is unclear and why.\n"
    "Format your response as a numbered list."
)

REVISE_PROMPT = (
    "You are a skilled writer. You will receive a task, a draft answer, and a critique.\n"
    "Produce a revised answer that directly addresses every point in the critique. "
    "Your revision should be complete and self-contained — do not reference the critique in your output."
)


# ---------------------------------------------------------------------------
# TODO 1: Implement draft(task_prompt).
#         Generate an initial answer to the task. Use temperature=0.7 so the
#         output is natural but not overly conservative.
#         Return the text of the draft.
# ---------------------------------------------------------------------------
def draft(task_prompt: str) -> str:
    # Get the default provider, make one chat call with a list[ChatMessage]
    # holding just the task_prompt as a "user" message, using
    # ChatOptions(temperature=0.7), and return the reply text.
    raise NotImplementedError("TODO: implement draft()")


# ---------------------------------------------------------------------------
# TODO 2: Implement critique(task_prompt, draft_text).
#         Send a message with the critique system prompt + the task + the draft.
#         Return the critique text.
#         Temperature=0 keeps the critique factual and repeatable.
# ---------------------------------------------------------------------------
def critique(task_prompt: str, draft_text: str) -> str:
    # Build a list[ChatMessage]: the CRITIQUE_PROMPT constant as the "system"
    # message, and a "user" message that presents both the task and the draft
    # answer (label each so the model can tell them apart). Call llm.chat with
    # ChatOptions(temperature=0) and return the reply text.
    raise NotImplementedError("TODO: implement critique()")


# ---------------------------------------------------------------------------
# TODO 3: Implement revise(task_prompt, draft_text, critique_text).
#         Send a message with the revise system prompt + task + draft + critique.
#         Return the revised text.
# ---------------------------------------------------------------------------
def revise(task_prompt: str, draft_text: str, critique_text: str) -> str:
    # Build a list[ChatMessage]: the REVISE_PROMPT constant as the "system"
    # message, and a "user" message that supplies all three pieces the reviser
    # needs — the task, the draft answer, and the critique (label each). Call
    # llm.chat with a low-but-nonzero temperature (e.g. ChatOptions(
    # temperature=0.3)) and return the reply text.
    raise NotImplementedError("TODO: implement revise()")


# ---------------------------------------------------------------------------
# TODO 4: Implement compare(original, final).
#         A simple heuristic comparison: length change, unique words added,
#         etc. Return a dict with at least {"length_delta": int, "changed": bool}.
#         This is a stub — replace with an LLM-as-judge call if you like.
# ---------------------------------------------------------------------------
def compare(original: str, final: str) -> dict:
    # TODO: implement
    length_delta = len(final) - len(original)
    original_words = set(original.lower().split())
    final_words = set(final.lower().split())
    new_words = final_words - original_words
    return {
        "length_delta": length_delta,
        "new_unique_words": len(new_words),
        "changed": original.strip() != final.strip(),
    }


def divider(label: str) -> None:
    print(f"\n{'='*10} {label} {'='*10}")


def main() -> None:
    print("=== Task 3: Self-Refine / Reflection ===\n")

    for task in TASKS:
        print(f"\nTask [{task['id']}]: {task['prompt']}\n")

        # -------------------------------------------------------------------------
        # TODO 5: Run the draft → critique → revise loop.
        #         1. Generate initial draft.
        #         2. For each iteration: critique the current draft, revise it.
        #         3. Print each step so the reader can see the improvement.
        # -------------------------------------------------------------------------

        try:
            # Step 1: initial draft
            divider("DRAFT")
            current = draft(task["prompt"])
            original = current
            print(current)

            # Iteration loop
            for iteration in range(1, MAX_ITERATIONS + 1):
                divider(f"CRITIQUE (iteration {iteration})")
                crit = critique(task["prompt"], current)
                print(crit)

                divider(f"REVISION (iteration {iteration})")
                current = revise(task["prompt"], current, crit)
                print(current)

            # -------------------------------------------------------------------------
            # TODO 6: Print a comparison between the original draft and final revision.
            # -------------------------------------------------------------------------
            divider("COMPARISON")
            stats = compare(original, current)
            print(f"  Length change     : {stats['length_delta']:+d} chars")
            print(f"  New unique words  : {stats['new_unique_words']}")
            print(f"  Changed at all?   : {stats['changed']}")

        except NotImplementedError as e:
            print(f"  {e}")

        print()

    print(
        "Observation: self-refinement tends to add specificity and fix factual gaps.\n"
        "Two iterations usually capture most of the gain."
    )


if __name__ == "__main__":
    main()
